### DISCUSSION TITLE
Signal Handling

### DISCUSSION THREAD

**Discussion by nealef** [rpm-software-management/rpm#3150] [Q&A] [answered]
If I establish signal handlers prior to invoking an RPM function then when I return from that function the signal handlers are no longer "active". For example,
```
import rpm
import sys
import signal

def signal_handler(__signo, __stack):
    print("Signaled")
    sys.exit(1)

signal.signal(signal.SIGTERM, signal_handler)

ti = rpm.TransactionSet()
mi = ti.dbMatch("name", "docker-ce")
for h in mi:
    print(h)
    break

signal.signal(signal.SIGTERM, signal_handler)

while True:
    pass
```
If I move the `signal.signal(...` to before the `while` loop then it works as expected.

I installed `rpm` via `pip3 install rpm`.

Is there a call I should be making to clean up the RPM environment that would restore my signal handler?

Thanks... Neale

**Comment by pmatilai** [comment, marked as answer]
Sorry for the late response.

The question is somewhat rpm version dependent: rpm versions >= 4.18 do not mess with signal delivery, but older versions indeed do due to Berkeley DB requirements. Various signals get hijacked by rpm at the time of rpmdb open, and should be returned to previous state when the db and all cursors are closed. What it comes down to is basically ensuring match iterators (mi in your example) or transaction sets linger around.

The former is best ensured by avoiding placing them in variables in the first place, eg use this instead:
```
for h in ti.dbMatch("name", "docker-ce"):
    print(h)
```

The database connection is hidden behind the ts object, it'll get garbage collected and closed when it goes out of scope but if you can manually achieve that with either `ti.closeDB()` (but then you'll need to manually reopen the db if you need it again) or just `del ti`.

**Reply by nealef** [reply]
Thanks. I found the `del`'ing the object did the trick but wasn't sure if I'd missed a call that would clean things up.
